### ISSUE TITLE
A potential issue on node value initialization

### ISSUE THREAD

**Issue by Zhiyuan-Wu** [Zhiyuan-Wu/alpha-zero-gobang#1]
In original AlphaGo Zero paper, the author claimed that they initialize a leaf node with `Q(s, a)=0`, however, I guess this maybe problemtic and cause weak play under disadvantag state.

Suppose a leaf node is initialize with `Q(s, a)=0`. When another search comes to this node, PUCT rule will basically select a child `c` of `s` according to prior policy evaluated by the network, as all child of `s` has equal value 0. This behavior is well expected.

However,  **In case of a disadvantag state**, the network will estimate a negative value for `c`, which will cause `Q(s, a)<0` for the action `a` lead to `c`. This will significantly reduce the  PUCT score of child `c` and prevent next few searches visiting it. As a result, we have to visit all child of `s` to set their value before we move deeper to any of them.

This would be problemtic when the total search number is restricted. Perhaps the root node would have enough search chance to walk through all child node, but deeper nodes would not. This may lead to incorrect value estimation for subtrees and finally affect the quality of root visit result.

Here is (probably) an evidence. These images plot the search result for current state, with estimated value (normalized to 0-1 as winrate) above and visit count below for some possible moves. These information will only appear when that child is once visited. The first image is result for black move, and sencond for white move. We can see that black is at a disadvantage state, and search result covers entire board. Some of move (e.g. A15) is ridiculous but still got a 24% estimated winrate. This perhaps waste too much search, mislead to a wrong decision H8 instead of a better choice E4.

<a href="https://ibb.co/nzYX0KP"><img src="https://i.ibb.co/5BSQ9JK/20230802224706.png" alt="20230802224706" border="0" width=40%></a>
<a href="https://ibb.co/0DYvdb2"><img src="https://i.ibb.co/j63s1Xf/20230802224845.png" alt="20230802224845" border="0" width=40%></a>

On the other hand, we in fact expect continuals behavior. The first a few visits should follow prior policy just like the first visit, but not suddenly switch to walk-through-all behavior. This issue could be the reason why current model plays very poor white.

A larger PUCT coefficient may force search better follow the prior policy at small visit number. The choice of setting it to 1 as suragnair's project is not a good choice. But this does not solve the problem, we need a more clever way to fix this potential problem.

**Comment by Zhiyuan-Wu** [comment]
Answer from ChatGPT:

The issue you've described is indeed a valid concern. Initializing the leaf node with Q(s, a) = 0 can lead to problems in disadvantagous states, as the negative estimated values for children can significantly reduce their PUCT scores and prevent them from being explored further. This could lead to incorrect value estimations for subtrees and potentially affect the overall performance of the search algorithm.

One way to address this problem is to use a more sophisticated initialization strategy for the leaf nodes. Instead of setting Q(s, a) to a fixed value such as 0, you could consider using a learned initialization. For example, you can use the output of the neural network as an initialization value for Q(s, a) when the node is expanded. This allows the search algorithm to leverage the initial evaluations provided by the neural network, which are likely to be more accurate than a fixed value like 0.

Another approach is to use a progressive widening strategy during the search. Progressive widening dynamically increases the number of allowable child actions as the number of visits to a node increases. This means that during early visits to a node, the search will be more biased towards exploring actions according to the prior policy. As the number of visits increases, the search will gradually shift towards a more exhaustive exploration of the actions. This allows the search to maintain the desired balance between exploration and exploitation throughout the search process.

You can also experiment with different values for the PUCT coefficient (exploration-exploitation trade-off parameter) to control the balance between prior policy guidance and exploration. A higher PUCT coefficient will place more emphasis on the prior policy, while a lower value will encourage more exploration. Finding the right balance for your specific problem and search setting can be crucial for achieving strong performance.

In summary, some potential solutions to address the problem of weak play under disadvantageous states in the AlphaGo Zero-like algorithm include:

- Using a learned initialization for Q(s, a) based on the neural network's output.
- Employing progressive widening to dynamically adjust the number of allowable child actions during the search.
- Experimenting with different values for the PUCT coefficient to control exploration-exploitation trade-off.

By carefully selecting and combining these techniques, you may be able to mitigate the issue and improve the overall performance of the search algorithm.
